### PR TITLE
Wait for next reconciliation after sync

### PR DESCRIPTION
**Pull request by sboschman**
rough idea for Slack topic https://argoproj.slack.com/archives/CSPC6G5JQ/p1582708685014500

### CODE REVIEW

File: controller/controller.go [hunk 1]
```diff
@@ -232,6 +232,52 @@ func (c *notificationController) processApp(app *unstructured.Unstructured, logE
 	return nil
 }
 
+// Checks if the application SyncStatus has been refreshed by Argo CD after an operation has completed
+func (c *notificationController) isAppSyncStatusRefreshed(app *unstructured.Unstructured, logEntry *log.Entry) bool {
```
Comment (alexmt) [inline controller/controller.go]: Any chance we have this check inside of the trigger [condition](https://github.com/argoproj-labs/argocd-notifications/blob/master/builtin/triggers.go#L29)? 

The condition has access to the  `time.Parse` function.  This should be enough to verify that `reconciledAt ` is after sync completion. E.g.

```
app.status.operationState.phase in ['Succeeded'] && time.Parse(app.status.operationState.finishedAt).Before(time.Parse(app.status.reconciledAt))
```

The condition might fail and it is treated as false, so it is ok if `finishedAt` or `reconciledAt` are empty and `time.Parse` fails.

Comment (sboschman) [reply controller/controller.go]: I guess the `on-sync-status-unknown` and `on-sync-failed` triggers will need this check as well (and most likely `on-health-degraded` too).

My objection against solving it in the trigger condition itself is that anyone creating a user-defined trigger has to add this check as well (or perhaps not if they don't use the sync status in their template). This probably gets rather confusing. It sort of creates a relation between trigger condition and template, i.e. the condition has to know which variables are being used in the template. What is your opinion regarding this?

How about adding an [extra condition](https://github.com/argoproj-labs/argocd-notifications/blob/master/triggers/triggers.go#L153) evaluation here ? This condition is the above check and is defined by argocd-notifcations, but isn't part of 'every' trigger condition. Let me see if I can code something together to discuss further.

Comment (sboschman) [reply controller/controller.go]: interesting, the following condition works:
`time.Parse(app.status.operationState.finishedAt).After(time.Now())`
but this one does not:
`time.Parse(app.status.operationState.finishedAt).After(time.Parse(app.status.observedAt))`

`reflect: Call using zero Value argument` at the app in the second time.Parse

Comment (sboschman) [reply controller/controller.go]: the (built-in/user) defined trigger condition is wrapped with a sync status check:
https://github.com/argoproj-labs/argocd-notifications/pull/51/files#diff-56e03da81512d7963e3a61119bcbf36cR153

Comment (alexmt) [reply controller/controller.go]: I did not realize that check is required for all triggers. After thinking more about it I agree with you. I think we should just skip app reconciliation if sync operation just completed and reconciledAt is less than sync operation completion.

File: controller/controller.go [hunk 2]
```diff
@@ -232,6 +232,52 @@ func (c *notificationController) processApp(app *unstructured.Unstructured, logE
...
+	phase, ok, err := unstructured.NestedString(app.Object, "status", "operationState", "phase")
+	if !ok || err != nil {
+		logEntry.Debug("No OperationPhase found, SyncStatus is assumed to be up-to-date")
+		return true
+	}
+	switch phase {
+	case "Failed", "Error", "Succeeded":
+		finishedAtRaw, ok, err := unstructured.NestedString(app.Object, "status", "operationState", "finishedAt")
+		if !ok || err != nil {
+			logEntry.Debugf("No FinishedAt found for completed phase '%s', SyncStatus is assumed to be out-of-date", phase)
```
Comment (alexmt) [inline controller/controller.go]: I think notifications should be skipped only for a short period of time after sync operation completion and before the next reconciliation. So if sync is in progress and finishedAt is missing we still need to send notifications.

Comment (sboschman) [reply controller/controller.go]: Could you explain in more depth what you mean please?

As far as I understand the argo-cd code, the phase and finishedAt are set [at the same time](https://github.com/argoproj/argo-cd/blob/master/controller/appcontroller.go#L733) for [completed phases](https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/types.go#L484).

Of course `unstructured.NestedString` does not know this and we have to check the `ok` and `err`. Maybe the debug msg should be different, but it is a (in)famous 'this should never happen' msg I guess.

If the sync is in progress the phase should be `Running`, and we should not hit the finishedAt check at all. For the phases running and terminating we just assume the SyncStatus is up to date and send out notifications right away. Only for completed phases (failed, error and succeeded) we delay sending the notification till argo-cd has updated the SyncStatus by checking the observedAt and reconciledAt timestamps.

Comment (alexmt) [reply controller/controller.go]: Oh, you are right, I've missed that `finishedAt` is verified after checking the operation phase. 👍

**Pull request merged**